jax/_src/random.py

```diff
@@ -2627,6 +2627,65 @@ def binomial(
 batching.defvectorized(random_clone_p)
 mlir.register_lowering(random_clone_p, lambda _, k: [k])
 
+
+def multinomial(
+    key: Array,
+    n: RealArray,
+    p: RealArray,
+    axis: int = -1,
+    shape: Shape | None = None,
+):
+  r"""Sample from a multinomial distribution.
+
+  The probability mass function is
+
+  .. math::
+      f(x;n,p) = \frac{n!}{x_1! \ldots x_k!} p_1^{x_1} \ldots p_k^{x_k}
+
+  Args:
+    key: a PRNG key used as the random key.
+    n: a float array-like representing the number of trials.
+    p: a float array-like representing the probabilities of each outcome.
+    axis: axis along which probabilities are defined for each outcome.
+    shape: optional, a tuple of nonnegative integers specifying the result
+      shape. Must be broadcast-compatible with ``n`` and ``p``.
+
+  Returns:
+    An array of counts for each outcome.
+  """
+
+  key, _ = _check_prng_key("multinomial", key)
+  check_arraylike("multinomial", n, p)
+
+  if shape is not None:
+    p = jnp.broadcast_to(p, shape)
```

Comment:
I'm a bit confused by this broadcast: `shape` seems to relate to `p.shape`, while below `shape` seems to relate to `p.shape[1:]`.

Could we streamline this logic so all `shape` handling is in a single block, to make it easier to follow and also more efficient (e.g. currently `p` is broadcast twice).